### PR TITLE
plan, statistics: decouple `Seletivity` from `statistics.Table`.

### PR DESCRIPTION
**Pull request by winoros**
So we can do further optimizations.
PTAL @zz-jason @lamxTyler @XuHuaiyu

### CODE REVIEW

File: statistics/selectivity.go [hunk 1]
```diff
@@ -34,7 +34,9 @@ type exprSet struct {
 	// mask is a bit pattern whose ith bit will indicate whether the ith expression is covered by this index/column.
 	mask int64
 	// ranges contains all the ranges we got.
-	ranges []*ranger.Range
+	ranges  []*ranger.Range
+	colHist *Column
+	idxHist *Index
```
Comment (lamxTyler) [inline statistics/selectivity.go]: Seems that it is not used?

File: statistics/selectivity.go [hunk 2]
```diff
@@ -44,77 +46,71 @@ const (
...
-			}
-			if mysql.HasPriKeyFlag(col.Info.Flag) {
-				if t.PKIsHandle {
-					return 1.0 / float64(t.Count)
+			existsCol[col.ColName.L] = true
+			if mysql.HasUniKeyFlag(col.RetType.Flag) {
+				if fun.FuncName.L != ast.NullEQ {
+					return 1.0 / float64(totalCount)
+				}
+				if mysql.HasNotNullFlag(col.RetType.Flag) {
```
Comment (lamxTyler) [inline statistics/selectivity.go]: `NullEQ` is not `IsNull`.

File: statistics/selectivity.go [hunk 3]
```diff
@@ -44,77 +46,71 @@ const (
 	colType
 )
 
-const unknownColumnID = math.MinInt64
-
 // getConstantColumnID receives two expressions and if one of them is column and another is constant, it returns the
 // ID of the column.
-func getConstantColumnID(e []expression.Expression) int64 {
+func getConstantColumnID(e []expression.Expression) *expression.Column {
```
Comment (zz-jason) [inline statistics/selectivity.go]: s/getConstantColumnID/getConstantColumn/

File: statistics/selectivity.go [hunk 4]
```diff
@@ -44,77 +46,71 @@ const (
 	colType
 )
 
-const unknownColumnID = math.MinInt64
-
 // getConstantColumnID receives two expressions and if one of them is column and another is constant, it returns the
```
Comment (zz-jason) [inline statistics/selectivity.go]: update comment

File: statistics/selectivity.go [hunk 5]
```diff
@@ -44,77 +46,71 @@ const (
...
 	}
-	return unknownColumnID
+	return nil
 }
 
-func pseudoSelectivity(t *Table, exprs []expression.Expression) float64 {
+func pseudoSelectivity(exprs []expression.Expression, totalCount int64, idxHists map[int64]*Index) float64 {
 	minFactor := selectionFactor
-	uniqueCol := make(map[string]bool)
+	existsCol := make(map[string]bool)
```
Comment (zz-jason) [inline statistics/selectivity.go]: s/existsCol/colExists/

**Comment by winoros** [comment]
@zz-jason @lamxTyler PTAL

File: statistics/selectivity.go [hunk 6]
```diff
@@ -46,9 +44,9 @@ const (
 	colType
 )
 
-// getConstantColumnID receives two expressions and if one of them is column and another is constant, it returns the
+// getConstantColumn receives two expressions and if one of them is column and another is constant, it returns the
 // ID of the column.
```
Comment (zz-jason) [inline statistics/selectivity.go]: returns the column

**Comment by winoros** [comment]
I'll choose another way to do this.

**Pull request closed without merging**